File: es54x/src/main/java/tech/beshu/ror/es/IndexLevelActionFilter.java
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */

package tech.beshu.ror.es;

import org.elasticsearch.ElasticsearchException;
import org.elasticsearch.ResourceNotFoundException;
import org.elasticsearch.action.ActionListener;
import org.elasticsearch.action.ActionRequest;
import org.elasticsearch.action.ActionResponse;
import org.elasticsearch.action.support.ActionFilter;
import org.elasticsearch.action.support.ActionFilterChain;
import org.elasticsearch.client.node.NodeClient;
import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
import org.elasticsearch.cluster.service.ClusterService;
import org.elasticsearch.common.component.AbstractComponent;
import org.elasticsearch.common.inject.Inject;
import org.elasticsearch.common.inject.Singleton;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.common.xcontent.ToXContent;
import org.elasticsearch.common.xcontent.XContentBuilder;
import org.elasticsearch.common.xcontent.json.JsonXContent;
import org.elasticsearch.rest.BytesRestResponse;
import org.elasticsearch.rest.RestChannel;
import org.elasticsearch.rest.RestStatus;
import org.elasticsearch.tasks.Task;
import org.elasticsearch.threadpool.ThreadPool;
import org.elasticsearch.transport.TransportService;
import tech.beshu.ror.acl.ACL;
import tech.beshu.ror.commons.BasicSettings;
import tech.beshu.ror.configuration.ReloadableSettings;
import tech.beshu.ror.es.actionlisteners.ACLActionListener;
import tech.beshu.ror.es.requestcontext.RequestContextImpl;
import tech.beshu.ror.settings.RorSettings;
import tech.beshu.ror.commons.shims.ACLHandler;
import tech.beshu.ror.commons.shims.ESContext;
import tech.beshu.ror.commons.shims.LoggerShim;

import java.io.IOException;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;


/**
 * Created by sscarduzio on 19/12/2015.
 */
@Singleton
public class IndexLevelActionFilter extends AbstractComponent implements ActionFilter, Consumer<BasicSettings> {

  private final ThreadPool threadPool;
  private final ClusterService clusterService;

  private final AtomicReference<Optional<ACL>> acl;
  private final ESContext context;
  private final ReloadableSettings reloadableSettings;
  private final NodeClient client;
  private final LoggerShim logger;
  private final IndexNameExpressionResolver indexResolver;

  @Inject
  public IndexLevelActionFilter(Settings settings,
                                ClusterService clusterService,
                                TransportService transportService,
                                NodeClient client,
                                ThreadPool threadPool,
                                ReloadableSettingsImpl reloadableSettings,
                                IndexNameExpressionResolver indexResolver
  )
    throws IOException {
    super(settings);
    this.context = new ESContextImpl();
    this.logger = context.logger(getClass());

    this.reloadableSettings = reloadableSettings;
    this.clusterService = clusterService;
    this.indexResolver = indexResolver;
    this.threadPool = threadPool;
    this.acl = new AtomicReference<>(Optional.empty());
    this.client = client;
    this.reloadableSettings.addListener(this);

    new TaskManagerWrapper(settings).injectIntoTransportService(transportService, logger);

    logger.info("Readonly REST plugin was loaded...");
  }

  @Override
  public void accept(BasicSettings bsettings) {
    if (bsettings.isEnabled()) {
      try {
        AuditSinkImpl audit = new AuditSinkImpl(client, bsettings);
        ACL acl = new ACL(bsettings, this.context, audit);
        this.acl.set(Optional.of(acl));
        logger.info("Configuration reloaded - ReadonlyREST enabled");
      } catch (Exception ex) {
        logger.error("Cannot configure ReadonlyREST plugin", ex);
      }
    }
    else {
      this.acl.set(Optional.empty());
      logger.info("Configuration reloaded - ReadonlyREST disabled");
    }
  }

  @Override
  public int order() {
    return 0;
  }

  @Override
  public <Request extends ActionRequest, Response extends ActionResponse> void apply(Task task,
                                                                                     String action,
                                                                                     Request request,
                                                                                     ActionListener<Response> listener,
                                                                                     ActionFilterChain<Request, Response> chain) {
    Optional<ACL> acl = this.acl.get();
    if (acl.isPresent()) {
      handleRequest(acl.get(), task, action, request, listener, chain);
    }
    else {
      chain.proceed(task, action, request, listener);
    }
  }

  private <Request extends ActionRequest, Response extends ActionResponse> void handleRequest(ACL acl,
                                                                                              Task task,
                                                                                              String action,
                                                                                              Request request,
                                                                                              ActionListener<Response> listener,
                                                                                              ActionFilterChain<Request, Response> chain) {
    RestChannel channel = ThreadRepo.channel.get();
    boolean chanNull = channel == null;
    boolean reqNull = channel == null ? true : channel.request() == null;
    if (ACL.shouldSkipACL(chanNull, reqNull)) {
      chain.proceed(task, action, request, listener);
      return;
    }

    RequestContextImpl rc = new RequestContextImpl(channel.request(), action, request, clusterService, threadPool, context, indexResolver);

    acl.check(rc, new ACLHandler() {
      @Override
      public void onForbidden() {
        sendNotAuthResponse(channel, acl.getSettings());
      }

      @Override
      public void onAllow(Object blockExitResult) {
        boolean hasProceeded = false;
        try {
          @SuppressWarnings("unchecked")
          ActionListener<Response> aclActionListener = (ActionListener<Response>) new ACLActionListener(
            request, (ActionListener<ActionResponse>) listener, rc, blockExitResult, context, acl
          );
          chain.proceed(task, action, request, aclActionListener);
          hasProceeded = true;
          return;
        } catch (Throwable e) {
          e.printStackTrace();
        }
        if (!hasProceeded) {
          chain.proceed(task, action, request, listener);
        }
      }

      @Override
      public boolean isNotFound(Throwable throwable) {
        return throwable.getCause() instanceof ResourceNotFoundException;
      }

      @Override
      public void onNotFound(Throwable throwable) {
        sendNotFound((ResourceNotFoundException) throwable.getCause(), channel);
      }

      @Override
      public void onErrored(Throwable t) {
        sendNotAuthResponse(channel, acl.getSettings());
      }
    });

  }

  private void sendNotAuthResponse(RestChannel channel, RorSettings rorSettings) {
    BytesRestResponse resp;
    boolean doesRequirePassword = acl.get().map(ACL::doesRequirePassword).orElse(false);
    if (doesRequirePassword) {
      resp = new BytesRestResponse(RestStatus.UNAUTHORIZED, BytesRestResponse.TEXT_CONTENT_TYPE, rorSettings.getForbiddenMessage());
      logger.debug("Sending login prompt header...");
      resp.addHeader("WWW-Authenticate", "Basic");
    }
    else {
      resp = new BytesRestResponse(RestStatus.FORBIDDEN, BytesRestResponse.TEXT_CONTENT_TYPE, rorSettings.getForbiddenMessage());
    }

    channel.sendResponse(resp);
  }

  private void sendNotFound(ResourceNotFoundException e, RestChannel channel) {
    try {
      XContentBuilder b = JsonXContent.contentBuilder();
      b.startObject();
      ElasticsearchException.generateFailureXContent(b, ToXContent.EMPTY_PARAMS, e, true);
      b.endObject();
      BytesRestResponse resp;
      resp = new BytesRestResponse(RestStatus.NOT_FOUND, "application/json", b.string());
      channel.sendResponse(resp);
    } catch (Exception e1) {
      e1.printStackTrace();
    }
  }

}

File: es79x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.es

import monix.execution.atomic.Atomic
import org.apache.logging.log4j.scala.Logging
import org.elasticsearch.action.support.{ActionFilter, ActionFilterChain}
import org.elasticsearch.action.{ActionListener, ActionRequest, ActionResponse}
import org.elasticsearch.client.node.NodeClient
import org.elasticsearch.cluster.service.ClusterService
import org.elasticsearch.env.Environment
import org.elasticsearch.repositories.RepositoriesService
import org.elasticsearch.tasks.Task
import org.elasticsearch.threadpool.ThreadPool
import org.elasticsearch.transport.RemoteClusterService
import tech.beshu.ror.accesscontrol.domain.AuditCluster
import tech.beshu.ror.accesscontrol.matchers.UniqueIdentifierGenerator
import tech.beshu.ror.boot.ReadonlyRest.{AuditSinkCreator, RorMode}
import tech.beshu.ror.boot.RorSchedulers.Implicits.mainScheduler
import tech.beshu.ror.boot._
import tech.beshu.ror.boot.engines.Engines
import tech.beshu.ror.es.handler.AclAwareRequestFilter
import tech.beshu.ror.es.handler.AclAwareRequestFilter.{EsChain, EsContext}
import tech.beshu.ror.es.handler.response.ForbiddenResponse.{createRorNotReadyYetResponse, createRorStartingFailureResponse, createTestSettingsNotConfiguredResponse}
import tech.beshu.ror.es.services.{EsAuditSinkService, EsIndexJsonContentService, EsServerBasedRorClusterService, HighLevelClientAuditSinkService}
import tech.beshu.ror.es.utils.ThreadRepo
import tech.beshu.ror.exceptions.StartingFailureException
import tech.beshu.ror.providers.{EnvVarsProvider, PropertiesProvider}
import tech.beshu.ror.utils.AccessControllerHelper._
import tech.beshu.ror.utils.{JavaConverters, RorInstanceSupplier}

import java.time.Clock
import java.util.function.Supplier
import scala.language.postfixOps

class IndexLevelActionFilter(nodeName: String,
                             clusterService: ClusterService,
                             client: NodeClient,
                             threadPool: ThreadPool,
                             env: Environment,
                             remoteClusterServiceSupplier: Supplier[Option[RemoteClusterService]],
                             repositoriesServiceSupplier: Supplier[Option[RepositoriesService]],
                             esInitListener: EsInitListener)
                            (implicit envVarsProvider: EnvVarsProvider,
                             propertiesProvider: PropertiesProvider,
                             generator: UniqueIdentifierGenerator)
  extends ActionFilter with Logging {

  private implicit val clock: Clock = Clock.systemUTC()

  private val ror = ReadonlyRest.create(
    RorMode.Plugin,
    new EsIndexJsonContentService(client),
    auditSinkCreator,
    env.configFile
  )

  private val rorInstanceState: Atomic[RorInstanceStartingState] =
    Atomic(RorInstanceStartingState.Starting: RorInstanceStartingState)

  private val aclAwareRequestFilter = new AclAwareRequestFilter(
    new EsServerBasedRorClusterService(
      clusterService,
      remoteClusterServiceSupplier,
      repositoriesServiceSupplier,
      client,
      threadPool
    ),
    clusterService.getSettings,
    threadPool
  )

  private val startingTaskCancellable = startRorInstance()

  private def auditSinkCreator: AuditSinkCreator = {
    case AuditCluster.LocalAuditCluster =>
      new EsAuditSinkService(client, threadPool)
    case remote: AuditCluster.RemoteAuditCluster =>
      HighLevelClientAuditSinkService.create(remote)
  }

  override def order(): Int = 0

  def stop(): Unit = {
    startingTaskCancellable.cancel()
    rorInstanceState.get() match {
      case RorInstanceStartingState.Starting =>
      case RorInstanceStartingState.Started(instance) => instance.stop().runSyncUnsafe()
      case RorInstanceStartingState.NotStarted(_) =>
    }
  }

  override def apply[Request <: ActionRequest, Response <: ActionResponse](task: Task,
                                                                           action: String,
                                                                           request: Request,
                                                                           listener: ActionListener[Response],
                                                                           chain: ActionFilterChain[Request, Response]): Unit = {
    doPrivileged {
      proceed(
        task,
        action,
        request,
        listener.asInstanceOf[ActionListener[ActionResponse]],
        new EsChain(
          chain.asInstanceOf[ActionFilterChain[ActionRequest, ActionResponse]],
          threadPool
        )
      )
    }
  }

  private def proceed(task: Task,
                      action: String,
                      request: ActionRequest,
                      listener: ActionListener[ActionResponse],
                      chain: EsChain): Unit = {
    ThreadRepo.getRorRestChannel match {
      case None =>
        chain.continue(nodeName, task, action, request, listener)
      case Some(_) if action.startsWith("internal:") =>
        chain.continue(nodeName, task, action, request, listener)
      case Some(channel) =>
        proceedByRorEngine(
          EsContext(
            channel,
            nodeName,
            task,
            action,
            request,
            listener,
            chain,
            JavaConverters.flattenPair(threadPool.getThreadContext.getResponseHeaders).toSet
          )
        )
    }
  }

  private def proceedByRorEngine(esContext: EsContext): Unit = {
    rorInstanceState.get() match {
      case RorInstanceStartingState.Starting =>
        logger.warn(s"[${esContext.requestContextId}] Cannot handle the request ${esContext.channel.request().path()} because ReadonlyREST hasn't started yet")
        esContext.listener.onFailure(createRorNotReadyYetResponse())
      case RorInstanceStartingState.Started(instance) =>
        instance.engines match {
          case Some(engines) =>
            handleRequest(engines, esContext)
          case None =>
            logger.warn(s"[${esContext.requestContextId}] Cannot handle the request ${esContext.channel.request().path()} because ReadonlyREST hasn't started yet")
            esContext.listener.onFailure(createRorNotReadyYetResponse())
        }
      case RorInstanceStartingState.NotStarted(_) =>
        logger.error(s"[${esContext.requestContextId}] Cannot handle the ${esContext.channel.request().path()} request because ReadonlyREST failed to start")
        esContext.listener.onFailure(createRorStartingFailureResponse())
    }
  }

  private def handleRequest(engines: Engines, esContext: EsContext): Unit = {
    aclAwareRequestFilter
      .handle(engines, esContext)
      .runAsync {
        case Right(result) => handleResult(esContext, result)
        case Left(ex) => esContext.listener.onFailure(new Exception(ex))
      }
  }

  private def handleResult(esContext: EsContext, result: Either[AclAwareRequestFilter.Error, Unit]): Unit = result match {
    case Right(_) =>
    case Left(AclAwareRequestFilter.Error.TestEngineNotConfigured) =>
      esContext.listener.onFailure(createTestSettingsNotConfiguredResponse())
  }

  private def startRorInstance() = {
    val startResult = for {
      _ <- esInitListener.waitUntilReady
      result <- ror.start()
    } yield result
    startResult.runAsync {
      case Right(Right(instance)) =>
        RorInstanceSupplier.update(instance)
        rorInstanceState.set(RorInstanceStartingState.Started(instance))
      case Right(Left(failure)) =>
        val startingFailureException = StartingFailureException.from(failure)
        logger.error("ROR starting failure:", startingFailureException)
        rorInstanceState.set(RorInstanceStartingState.NotStarted(startingFailureException))
      case Left(ex) =>
        val startingFailureException = StartingFailureException.from(ex)
        logger.error("ROR starting failure:", startingFailureException)
        rorInstanceState.set(RorInstanceStartingState.NotStarted(StartingFailureException.from(startingFailureException)))
    }
  }
}

private sealed trait RorInstanceStartingState
private object RorInstanceStartingState {
  case object Starting extends RorInstanceStartingState
  final case class Started(instance: RorInstance) extends RorInstanceStartingState
  final case class NotStarted(cause: StartingFailureException) extends RorInstanceStartingState
}